### DISCUSSION TITLE
Necessity of use_ref

### DISCUSSION THREAD

**Discussion by Napam** [pyedifice/pyedifice#196] [General]
Just wondering about the necessity of `use_ref` when we can just use regular Python "ways"? Like, is it important that I use `use_ref` for some internal reasons, or is it just completely fine to just "assign variables"?

For example if I would want a reference to a label I can do this

```python
the_label: Label | None = None

with VBoxView():
     the_label = Label(...)
```

If I were to follow the use_ref way it would have been like this:

```python

label_ref = cast(Reference[Label], use_ref())

with VBoxView():
     Label(...).register_ref(ref)
```

But, using the "Python way" allows you to do more stuff, such as getting the references dynamic list elements:
```python
labels: list[Label] = []

with VBoxView():
     for i in range(n):
         labels.append(Label(...))
```

In case I would want a reference to a Layout element that requires a context (e.g. VScrollView) I could use the walrus operator introduced in Python 3.8:

```python
container: VScrollView | None = None

with (container := VScrollView()):
    ...
```

**Comment by jamesdbrock** [comment]
These two ways of getting a reference to the rendered element will be equivalent on the first render.

```python
label = Label(...)
```
and
```python
Label(...).register_ref(label_ref)
```
But on subsequent renders, we will re-use the Label instance from the first render and only update its *props* . So on subsequent renders the Label referred to by `label` will not be pointing to the rendered Label object. The Label referred to by `label_ref` will be pointing to the rendered Label object, which is the Label object from the first render.

Good question, thanks.

**Comment by jamesdbrock** [comment]
Tell me more about why you want to “reference to a Layout element that requires a context”? Maybe I can suggest a good approach for you.

**Comment by Napam** [comment]
> Tell me more about why you want to “reference to a Layout element that requires a context”? Maybe I can suggest a good approach for you.

@jamesdbrock 

I just wanted to illustrate that you could "capture" context stuff without using `use_ref`. 

In my specific case I have a `VScrollView` of `Image` widgets, and I want to programmatically scroll to a specific image. I basically have this:

```python
    input_img_container: VScrollView | None = None
    image_widgets: list[Image] = []

    async def scroll_to_selected():
        if not input_img_container or not input_img_container.underlying:
            return

        scroll_area = input_img_container.underlying

        selected_img = images_widgets[selected_idx]
        underlying = selected_img.underlying

        if not underlying:
            return

        scroll_area.ensureWidgetVisible(underlying)

    use_async(scroll_to_selected, [])
    
    .
    .
    .

            with (
                input_img_container := VScrollView(style={"width": 600, "align": "top"})
            ):
                for i, image_path in enumerate(image_paths):
                    if i == selected_idx:
                        img = InputImage(image_path, i, selected=True, width=560)
                    else:
                        img = InputImage(
                            image_path, i, on_click=select_image, width=560
                        )
                    image_widgets.append(img)

```

**Comment by jamesdbrock** [comment]
I think that what you have there will work, because

```python
use_async(scroll_to_selected, [])
```

will only be called after the first render.

After the first render, the `input_img_container` and the `img` will refer to the rendered elements.

But after later renders, the `input_img_container` and the `img` will not refer to the rendered elements. For that, you have to use a `Reference`

**Comment by Napam** [comment]
It works in my case :)

Ok, I start to understand more now; so, the edifice instances in the subsequent renders, are just "shell" instances that don't really do much? Like, they don't have any "meat"? It is the original references that actually has the handles to the underlying Qt elements?

Also, say that I would want to do "reference requiring logic" on a _dynamic_ list of elements after an arbitrary number of re-renders. Also assume that said the list of elements change over time. Is it possible in that case to get the refs of all the elements?

**Reply by jamesdbrock** [reply]
> the edifice instances in the subsequent renders, are just "shell" instances that don't really do much? Like, they don't have any "meat"? It is the original references that actually has the handles to the underlying Qt elements?

Yes.

> say that I would want to do "reference requiring logic" on a dynamic list of elements after an arbitrary number of re-renders. Also assume that said the list of elements change over time (the list itself changes). Is it possible in that case to get the refs of all the elements?

That's a bit tricky but I think you can do it. The [`use_effect`](https://pyedifice.github.io/stubs/edifice.use_effect.html) is called “after render and after the underlying Qt Widgets are updated.” So you are guaranteed to be able to to iterate over the `children` during a `use_effect`, and you know that at that time all of the `underlying` will be non-`None`.

Something like this:

```python

@component
def ParentComponent(self:Element):

  vbox_ref:Reference[VBoxView] = use_ref()

  def do_children_op():
    for child in vbox_ref().children:
      do_special_operation(child.underlying)

  use_effect(do_children_op, None) # Run after every render

  with VBoxView().register_ref(vbox_ref):
    for i in range(10):
      Label(text=str(i))
```
 
(EDITED)

**Reply by Napam** [reply]
I see. Thank you very much for explaining! I appreciate it :)
